sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java

```diff
@@ -173,41 +215,50 @@ public int getMaxFrameSize() {
-                final SessionHandler handler = handlerProvider.createSessionHandler(connectionId,
+                final SessionHandler sessionHandler = handlerProvider.createSessionHandler(connectionId,
                     getFullyQualifiedNamespace(), key, connectionOptions.getRetry().getTryTimeout());
                 final Session session = connection.session();
 
-                BaseHandler.setHandler(session, handler);
-                final AmqpSession amqpSession = createSession(key, session, handler);
+                BaseHandler.setHandler(session, sessionHandler);
+                final AmqpSession amqpSession = createSession(key, session, sessionHandler);
                 final Disposable subscription = amqpSession.getEndpointStates()
                     .subscribe(state -> {
                     }, error -> {
-                            logger.info("connectionId[{}] sessionName[{}]: Error occurred. Removing and disposing"
-                                    + " session.", connectionId, sessionName, error);
-                            removeSession(key);
-                        }, () -> {
-                            logger.info("connectionId[{}] sessionName[{}]: Complete. Removing and disposing session.",
-                                connectionId, sessionName);
-                            removeSession(key);
-                        });
+                        // If we were already disposing of the connection, the session would be removed.
+                        if (isDisposed.get()) {
+                            return;
+                        }
+
+                        logger.info("connectionId[{}] sessionName[{}]: Error occurred. Removing and disposing"
+                            + " session.", connectionId, sessionName, error);
+                        removeSession(key);
+                    }, () -> {
+                        // If we were already disposing of the connection, the session would be removed.
+                        if (isDisposed.get()) {
+                            return;
+                        }
+
+                        logger.verbose("connectionId[{}] sessionName[{}]: Complete. Removing and disposing session.",
+                            connectionId, sessionName);
+                        removeSession(key);
+                    });
 
                 return new SessionSubscription(amqpSession, subscription);
             });
 
-            return sessionSubscription.getSession();
-        });
+            return sessionSubscription;
+        }).flatMap(sessionSubscription -> {
+            final Mono<AmqpEndpointState> activeSession = sessionSubscription.getSession().getEndpointStates()
+                .filter(e -> e == AmqpEndpointState.ACTIVE)
+                .next()
+                .timeout(retryPolicy.getRetryOptions().getTryTimeout(), Mono.error(new AmqpException(true,
```

Comment:
We should add java doc in `AmqpRetryOptions.getTryTimeout` : how this is used as time out for connection to service bus. We have few git issues where user ask how this property will effect or how they should finetune it.